### PR TITLE
Add check to verify if outbox and sagas use same persistence

### LINKED ISSUE #5261: Improve exception when multiple persistences are configured
When a user configures multiple persistences, even if they are for different purposes, an exception is thrown. That exception is not clear about the root cause of the problem.

```
endpointConfiguration.UsePersistence<NHibernatePersistence>();
endpointConfiguration.UsePersistence<InMemoryPersistence,StorageType.Sagas>();
```

When run it blows up with
```
System.InvalidCastException occurred
  HResult=0x80004002
  Message=Unable to cast object of type 'NServiceBus.Persistence.NHibernate.NHibernateLazyNativeTransactionSynchronizedStorageSession' to type 'NServiceBus.InMemorySynchronizedStorageSession'.
  Source=NServiceBus.Core
  StackTrace:
   at NServiceBus.InMemorySagaPersister.Save(IContainSagaData sagaData, SagaCorrelationProperty correlationProperty, SynchronizedStorageSession session, ContextBag context) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Persistence\InMemory\SagaPersister\InMemorySagaPersister.cs:line 77
```
This situation should present the user with a meaningful (custom) exception that guides them to a solution.

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Fixes #5261

### CODE REVIEW

File: src/NServiceBus.Core.Tests/Persistence/PersistenceStartupTests.cs [hunk 1]
```diff
@@ -15,5 +18,37 @@ public void Should_return_false_when_checking_if_persistence_supports_storage_ty
 
             Assert.IsFalse(supported);
         }
+
+        [Test]
+        public void Should_prevent_using_different_persistence_for_sagas_and_outbox()
+        {
+            var config = new EndpointConfiguration("MyEndpoint");
+            config.UsePersistence<InMemoryPersistence>();
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Persistence/PersistenceStartupTests.cs]: is there a need for this line at all?

File: src/NServiceBus.Core.Tests/Persistence/PersistenceStorageMergerTests.cs [hunk 2]
```diff
@@ -68,4 +54,29 @@ public FakePersistence()
...
+        [Test]
+        public void Should_not_use_other_supported_storages()
+        {
+            var config = new EndpointConfiguration("MyEndpoint");
+            config.UsePersistence<FakePersistence, StorageType.Sagas>();
+            var persistences = config.Settings.Get<List<EnabledPersistence>>("PersistenceDefinitions");
+
+            var resultedEnabledPersistences = PersistenceStorageMerger.Merge(persistences, config.Settings);
+
+            Assert.IsTrue(!resultedEnabledPersistences.Any(p => p.SelectedStorages.Contains(typeof(StorageType.Subscriptions))));
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Persistence/PersistenceStorageMergerTests.cs]:
```suggestion
            Assert.IsFalse(resultedEnabledPersistences.Any(p => p.SelectedStorages.Contains(typeof(StorageType.Subscriptions))));
```

would this be easier to read?

File: src/NServiceBus.Core/Persistence/PersistenceStartup.cs [hunk 3]
```diff
@@ -45,6 +48,19 @@ public void Run(SettingsHolder settings)
...
+        static void ValidateSagaAndOutboxPersistence(List<EnabledPersistence> enabledPersistences)
+        {
+            var sagaPersisterType = enabledPersistences.FirstOrDefault(p => p.SelectedStorages.Contains(typeof(StorageType.Sagas)));
+            var outboxPersisterType = enabledPersistences.FirstOrDefault(p => p.SelectedStorages.Contains(typeof(StorageType.Outbox)));
+
+            if (sagaPersisterType != null 
+                && outboxPersisterType != null
+                && sagaPersisterType.DefinitionType != outboxPersisterType.DefinitionType)
+            {
+                throw new Exception($"Sagas and Outbox need to use the same type of persistence. Saga is configured to use {sagaPersisterType.DefinitionType.Name}. Outbox is configured to use {outboxPersisterType.DefinitionType.Name}.");
```
Comment (timbussmann) [inline src/NServiceBus.Core/Persistence/PersistenceStartup.cs]:
```suggestion
                throw new Exception($"Sagas and the Outbox need to use the same type of persistence. Saga persistence is configured to use {sagaPersisterType.DefinitionType.Name}. Outbox persistence is configured to use {outboxPersisterType.DefinitionType.Name}.");
```

File: src/NServiceBus.Core/Persistence/PersistenceStartup.cs [hunk 4]
```diff
@@ -20,6 +21,8 @@ public void Run(SettingsHolder settings)
             var resultingSupportedStorages = new List<Type>();
             var diagnostics = new Dictionary<string, object>();
 
+            ValidateSagaAndOutboxPersistence(enabledPersistences);
```
Comment (SeanFeldman) [inline src/NServiceBus.Core/Persistence/PersistenceStartup.cs]: Nitpicking, but would it be better to indicate what the validation is about? E.g. `ValidateSagaAndOutboxPersistenceIsSame` or something among the lines? That way I don't have to dive into the implementation to understand what the method is supposed to do.

**Comment by SzymonPobiega** [comment]
@timbussmann @SeanFeldman I applied your suggestions

**Pull request merged**